File: rd-net/RdFramework/Base/RdReactiveBase.cs
using System;
using JetBrains.Annotations;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Impl;
using JetBrains.Serialization;

namespace JetBrains.Rd.Base
{
  public abstract class RdReactiveBase : RdBindableBase, IRdReactive
  {    
    internal static readonly ILog ourLogReceived = Protocol.Logger.GetSublogger("RECV");
    internal static readonly ILog ourLogSend = Protocol.Logger.GetSublogger("SEND");
    
    internal static LogWithLevel? ReceiveTrace => ourLogReceived.WhenTrace();
    internal static LogWithLevel? SendTrace => ourLogSend.WhenTrace();


    #region Assertion
    
    public bool Async { get; set; }

    [AssertionMethod]
    protected virtual void AssertThreading()
    {
      if (!Async && AllowBindCookie.IsBindNotAllowed && TryGetProto() is {} proto)
        proto.Scheduler.AssertThread(this);
    }

    public bool ValueCanBeNull { get; set; }

    protected internal void AssertNullability<T>(T value)
    {
      
      if ( //optimization for memory traffic 
        typeof(T).IsValueType || ValueCanBeNull || value != null) return;

      Assertion.Fail("Value is defined as not nullable: {0}", this);
    }

    [AssertionMethod]
    protected void AssertBound()
    {
      Assertion.Require(IsBound, "Not bound: {0}", this);
    }

    #endregion 


    #region Local change
    
    public bool IsLocalChange { get; protected set; }
    
    protected internal readonly struct LocalChangeCookie : IDisposable
    {
      private readonly RdReactiveBase myHost;
      private readonly FirstChanceExceptionInterceptor.ThreadLocalDebugInfo myDebugInfo;

      internal LocalChangeCookie(RdReactiveBase host)
      {
        (myHost = host).IsLocalChange = true;
        myDebugInfo = host.UsingDebugInfo();
      }      

      public void Dispose()
      {
        myHost.IsLocalChange = false;
        myDebugInfo.Dispose();
      }
    }

    protected internal FirstChanceExceptionInterceptor.ThreadLocalDebugInfo UsingDebugInfo()
    {
      return new FirstChanceExceptionInterceptor.ThreadLocalDebugInfo(this);
    }
    
    protected internal LocalChangeCookie UsingLocalChange()
    {
      if (Mode.IsAssertion) Assertion.Assert(!IsLocalChange, "!IsLocalChange: {0}", this);
      if (IsBound && !Async) AssertThreading();
      return new LocalChangeCookie(this);
    }
    #endregion


    #region From interface

    public void OnWireReceived(UnsafeReader reader, IRdWireableDispatchHelper dispatchHelper)
    {
      var proto = TryGetProto();
      if (proto == null || !TryGetSerializationContext(out var serializationCtx) || dispatchHelper.Lifetime.IsNotAlive)
      {
        ourLogReceived.Trace($"{this} is not bound. Message for ({dispatchHelper.RdId}) will not be processed");
        return;
      }
      
      OnWireReceived(proto, serializationCtx, reader, dispatchHelper);
    }
    
    public abstract void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader reader, IRdWireableDispatchHelper dispatchHelper);

    #endregion
  }
}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/base/RdReactiveBase.kt
package com.jetbrains.rd.framework.base

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.util.getLogger
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.isNotAlive
import com.jetbrains.rd.util.reactive.IScheduler
import com.jetbrains.rd.util.trace

abstract class RdReactiveBase : RdBindableBase(), IRdReactive {
    companion object {
        val logReceived = Protocol.sublogger("RECV")
        val logSend = Protocol.sublogger("SEND")

        val logAssert = getLogger<RdReactiveBase>()
    }

    private var masterOverriden : Boolean? = null
    var master : Boolean
        get() = masterOverriden ?: protocol?.isMaster ?: false
        set(value) { masterOverriden = value }

    //assertion
    override var async = false
    protected open fun assertThreading() {
        if (!async && AllowBindingCookie.isBindNotAllowed) {
            val proto = protocol ?: return
            proto.scheduler.assertThread(this)
        }
    }
    protected fun assertBound() {
        if (!isBound) { throw IllegalStateException("Not bound: $location") }
    }


    //local change
    var isLocalChange = false
        protected set(value) {
            field = value
        }

    internal fun <T> localChange(action: () -> T) : T {
        if (isBound && !async) assertThreading()

        require(!isLocalChange){ "!isLocalChange" }

        isLocalChange = true
        try {
            return action()
        } finally {
            isLocalChange = false
        }
    }

    final override fun onWireReceived(buffer: AbstractBuffer, dispatchHelper: IRdWireableDispatchHelper) {
        val proto = protocol
        val ctx = serializationContext
        if (proto == null || ctx == null || dispatchHelper.lifetime.isNotAlive) {
            logReceived.trace { "$this is not bound. Message for (${dispatchHelper.rdId} will not be processed" }
            return
        }

        return onWireReceived(proto, buffer, ctx, dispatchHelper)
    }

    abstract fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper)
}
